src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,191 @@
+        public override IEnumerable ChangedObjectCollection { get { return _items; } }
+
+        /// <inheritdoc/>
+        public void Add(IEdmChangedObject<TStructuralType> item)
+        {
+            _items.Add(item);
+        }
+
+       
+        /// <inheritdoc/>
+        public bool Contains(IEdmChangedObject<TStructuralType> item)
+        {
+            return _items.Contains(item);
+        }
+
+        /// <inheritdoc/>
+        public void CopyTo(IEdmChangedObject<TStructuralType>[] array, int arrayIndex)
+        {
+            _items.CopyTo(array, arrayIndex);
+        }
+
+        /// <inheritdoc/>
+        public bool IsReadOnly
+        {
+            get { return false; }
+        }
+
+        /// <inheritdoc/>
+        public bool Remove(IEdmChangedObject<TStructuralType> item)
+        {
+            return _items.Remove(item);
+        }
+
+        /// <inheritdoc/>
+        IEnumerator<IEdmChangedObject<TStructuralType>> IEnumerable<IEdmChangedObject<TStructuralType>>.GetEnumerator()
+        {
+            return _items.GetEnumerator();
+        }
+
+        /// <inheritdoc/>
+        protected override void InsertItem(int index, IEdmChangedObject item)
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            Contract.Assert(_item != null);
+
+            _items.Add(_item);
+        }
+
+        /// <inheritdoc/>
+        protected override void ClearItems()
+        {
+            _items.Clear();
+        }
+
+        /// <inheritdoc/>
+        protected override void RemoveItem(int index)
+        {
+            Contract.Assert(_items.Count > index);
+            
+            _items.RemoveAt(index);
+        }
+
+        /// <inheritdoc/>
+        protected override void SetItem(int index, IEdmChangedObject item)
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            Contract.Assert(_item != null);
+
+            _items[index] = _item;
+        }
+
+        /// <summary>
+        /// Copy changed values is an implementation of Patch
+        /// </summary>
+        /// <param name="original"></param>              
+        public void CopyChangedValues(ICollection<TStructuralType> original)
+        {
+            //Here we need to Find the key of the Type, then only we will be able to find from the collection that which item in the collection
+            //corresponds to the item in delta list(Edmchangedobjectcoll). For this we use somewhat the same logic used in 
+            //the method private static PropertyConfiguration GetKeyProperty(EntityTypeConfiguration entityType) in EntityKeyConvention class
+            //Once we find the key we use that to pick the corresponding item from the original collection (by comparing the value of the key, eg: Id)
+            Type type = original.First().GetType();
+            string key = GetKeyProperty(type.GetProperties(), type.Name);
```

Comment:
as discussed, just use the entityType passed into constructor and call entityType.Keys().  Note that a key can contain multiple properties; In the logic below, you will need to make sure that each key property matches.